zcash_primitives/src/transaction/builder.rs

```diff
@@ -0,0 +1,692 @@
+    /// [`OutputDescription`] in the transaction.
+    pub fn output_index(&self, n: usize) -> Option<usize> {
+        self.output_indices.get(n).map(|i| *i)
+    }
+}
+
+/// Generates a [`Transaction`] from its inputs and outputs.
+pub struct Builder<R: Rng> {
+    rng: R,
+    mtx: TransactionData,
+    fee: Amount,
+    anchor: Option<Fr>,
+    spends: Vec<SpendDescriptionInfo>,
+    outputs: Vec<SaplingOutput>,
+    change_address: Option<(OutgoingViewingKey, PaymentAddress<Bls12>)>,
+}
+
+impl Builder<OsRng> {
+    /// Creates a new `Builder` targeted for inclusion in the block with the given height,
+    /// using default values for general transaction fields and the default OS random.
+    ///
+    /// # Default values
+    ///
+    /// The expiry height will be set to the given height plus the default transaction
+    /// expiry delta (20 blocks).
+    ///
+    /// The fee will be set to the default fee (0.0001 ZEC).
+    pub fn new(height: u32) -> Self {
+        Builder::new_with_rng(
+            height,
+            OsRng::new().expect("should be able to construct RNG"),
+        )
+    }
+}
+
+impl<R: Rng> Builder<R> {
+    /// Creates a new `Builder` targeted for inclusion in the block with the given height
+    /// and randomness source, using default values for general transaction fields.
+    ///
+    /// # Default values
+    ///
+    /// The expiry height will be set to the given height plus the default transaction
+    /// expiry delta (20 blocks).
+    ///
+    /// The fee will be set to the default fee (0.0001 ZEC).
+    pub fn new_with_rng(height: u32, rng: R) -> Builder<R> {
+        let mut mtx = TransactionData::new();
+        mtx.expiry_height = height + DEFAULT_TX_EXPIRY_DELTA;
+
+        Builder {
+            rng,
+            mtx,
+            fee: DEFAULT_FEE,
+            anchor: None,
+            spends: vec![],
+            outputs: vec![],
+            change_address: None,
+        }
+    }
+
+    /// Adds a Sapling note to be spent in this transaction.
+    ///
+    /// Returns an error if the given witness does not have the same anchor as previous
+    /// witnesses, or has no path.
+    pub fn add_sapling_spend(
+        &mut self,
+        extsk: ExtendedSpendingKey,
+        diversifier: Diversifier,
+        note: Note<Bls12>,
+        witness: IncrementalWitness<Node>,
+    ) -> Result<(), Error> {
+        // Consistency check: all anchors must equal the first one
+        if let Some(anchor) = self.anchor {
+            let witness_root: Fr = witness.root().into();
+            if witness_root != anchor {
+                return Err(Error(ErrorKind::AnchorMismatch));
+            }
+        } else {
+            self.anchor = Some(witness.root().into())
+        }
+        let witness = witness.path().ok_or(Error(ErrorKind::InvalidWitness))?;
+
+        let alpha = Fs::rand(&mut self.rng);
+
+        self.mtx.value_balance.0 += note.value as i64;
```

Comment:
I'm worried about the potential for overflow here, since `note.value` can be set arbitrarily by the caller. I think we need to either error out when a `Note` is constructed with a value beyond `MAX_VALUE` or we need to error in every place where we do arithmetic like this.